### PR TITLE
upstream: update all subset load balancers to detect host health changes

### PR DESCRIPTION
**Pull request by zuercher**
Modifies the SubsetLoadBalancer to update healthy hosts for its subsets on update, even if there is no change to the overall set of hosts. By storing the predicate used to select hosts for a subset in the inner PrioritySubsetImpl we can update the subsets without having to recompute the predicate.

*Risk Level*: Medium

*Testing*: additional unit test that triggers the bug

*Fixes*: #2779 

Signed-off-by: Stephan Zuercher <stephan@turbinelabs.io>

### CODE REVIEW

**Comment by junr03** [comment]
@danielhochman can you take a first pass?

File: source/common/upstream/subset_lb.cc [hunk 1]
```diff
@@ -127,31 +127,34 @@ void SubsetLoadBalancer::updateFallbackSubset(uint32_t priority, const HostVecto
...
+  fallback_subset_->priority_subset_->update(priority, hosts_added, hosts_removed);
 }
 
 // Iterates over the added and removed hosts, looking up an LbSubsetEntryPtr for each. For every
-// unique LbSubsetEntryPtr found, it invokes cb with the LbSubsetEntryPtr, a HostPredicate that
-// selects hosts in the subset, and a flag indicating whether any hosts are being added.
+// unique LbSubsetEntryPtr found, it either invokes new_cb or update_cb depending on whether the
+// LbSubsetEntryPtr is already initialized (update_cb) or not (new_cb). In addition, update_cb is
+// invoked for any other other unmodified, but initialized LbSubsetEntryPtr to allow host health to
+// be updated.
```
Comment (htuch) [inline source/common/upstream/subset_lb.cc]: Nit: `any other unmodified and active, initialize LbSubsetEntryPtr, to allow`.

File: source/common/upstream/subset_lb.cc [hunk 2]
```diff
@@ -173,14 +176,29 @@ void SubsetLoadBalancer::processSubsets(
...
   }
+
+  forEachSubset(subsets_, [&](LbSubsetEntryPtr entry) {
+    if (subsets_modified.find(entry) != subsets_modified.end()) {
+      // Already handled due to hosts being added or removed.
+      return;
+    }
+
+    if (entry->initialized() && entry->active()) {
+      update_cb(entry);
```
Comment (htuch) [inline source/common/upstream/subset_lb.cc]: Do we double invoke `update_cb` here potentially (here and above)?

Comment (zuercher) [reply source/common/upstream/subset_lb.cc]: No, subsets_modified tracks all the subsets for which we've called either update_cb or new_cb.

Comment (zuercher) [reply source/common/upstream/subset_lb.cc]: I should clarify, it tracks the subsets for which new_cb or update_cb was called in block above. In this block, we see each subset once and only need to ignore those that were already updated (or just created).

**Pull request merged**